File: 2023/day1/src/part2.ts
import fs from 'fs';

function main() {
  const file = fs.readFileSync('./bin/input.txt').toString();

  const lines = file.split('\n');
  let sum = 0;

  enum words {
    'zero',
    'one',
    'two',
    'three',
    'four',
    'five',
    'six',
    'seven',
    'eight',
    'nine',
  }

  for (const line of lines) {
    // TODO:: fix this regex

    let reg = /\d|zero|one|two|three|four|five|six|seven|eight|nine/g;

    // TODO:: eightwo must be 82 not 88 or 22
    const res = line.match(reg);

    if (!res) {
      throw new Error('No match');
    }

    const f = isNaN(Number(res?.[0])) ? words[res?.[0]] : parseInt(res?.[0]);
    const l = isNaN(Number(res?.[res?.length - 1]))
      ? words[res?.[res?.length - 1]]
      : parseInt(res?.[res?.length - 1]);

    sum += parseInt(`${f}${l}`);
  }

  console.log(sum);

  return sum;
}

main();

File: 2023/day1/src/bin/part2.rs
fn main() {
    let content = include_str!("input.txt");
    let mut sum = 0;

    for line in content.lines() {
        let new_line = line
            .replace("one", "o1e")
            .replace("two", "t2o")
            .replace("three", "th3ee")
            .replace("four", "fo4r")
            .replace("five", "fi5e")
            .replace("six", "s6x")
            .replace("seven", "se7en")
            .replace("eight", "ei8ht")
            .replace("nine", "ni9e");

        let chars: Vec<char> = new_line.chars().collect();

        let mut i = 0;
        let mut j = new_line.len() - 1;

        while i <= j {
            if chars[i].is_numeric() && chars[j].is_numeric() {
                break;
            } else {
                if chars[i].is_alphabetic() {
                    i += 1;
                }

                if chars[j].is_alphabetic() {
                    j -= 1;
                }
            }
        }

        sum += format!("{}{}", chars[i], chars[j]).parse::<u32>().unwrap();
    }

    println!("{sum}");
}
